EIPS/eip-3171.md

```diff
@@ -0,0 +1,148 @@
+---
+eip: 3171
+title: CREATE3
+author: Moody Salem (@moodysalem)
+category: Core
+type: Standards Track
+status: Draft
+created: 2020-12-22
+---
+
+## Simple Summary
+
+This EIP is a fork of [EIP-1014](./eip-1014.md) (`CREATE2`) which adds an account creation opcode (`CREATE3`) that does
+not involve the init code hash. It is otherwise identical to `CREATE2` opcode.
+
+This opcode allows users to create accounts at an address independent of the bytecode deployed to the address.
+
+## Abstract
+
+TBD
+
+## Motivation
+
+The core motivations of the `CREATE3` opcode are the same as [EIP-1014](./eip-1014.md), i.e. deploying a smart contract at a
+deterministic address.
+
+This variation removes the hash of the init code from the preimage. This enables the address of contract created via `CREATE3`
+to be computed on-chain knowing only the deployer address and salt.
+
+It is possible to compute the address of a contract created with `CREATE2` without making any external calls. For example,
+Uniswap V2 [uses this](https://github.com/Uniswap/uniswap-v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol#L18)
+to save gas in computing the pair address from the Uniswap V2 Router. The gas cost of the `pairFor` method which
+computes the pair address for a given pair of tokens is only ~649 gas, while an external call + an SLOAD from mapping
+may cost upwards of 3k gas. A swap that is routed through multiple pairs requires multiple calls of this method.
+
+This is feasible only because the Uniswap V2 Pair does not have any constructor arguments, which means the init code hash
+can be hard coded in the library. However, the init code hash cannot be hard coded when the deployed smart contract has
+constructor arguments. In order to compute the init code hash, one must append the constructor arguments to the init code
+before computing the init code hash. Thus, in order to compute the address of an account created via `CREATE2`, it is 
+necessary to include the entire init code without the constructor arguments, and hash 
+`ceil(len(init_code_base + constructor_arguments) / 32)` words on-chain.
```

Comment:
It's not particularly pretty, but I believe another workaround to this problem is to have the init code pull the arguments from the creator:
```
interface F {
  function provideX() external returns (uint) {
  }

contract A {
   uint x;
   constructor() {
      x = F(msg.sender).provideX();
   }
}

contract Factory {
   uint x;
   function provideX() public returns (uint) {
      return x;
   }
   function deployA(uint initArg) public {
       x = initArg;
       new A();
       x = 0;
   }
}

Reply:
Thank you, I heard about this in the #allcoredevs channel of Discord and I will add it as a workaround